src/octrees.rs

```diff
@@ -0,0 +1,147 @@
+    }
+
+    /// Add a point from the top of the tree down. If this overfills the node,
+    /// subdivide it as necessary, up to the given max depth.
+    pub fn add_point(
+            &mut self, point: Vec3, color: Vec3, max_depth: u8) {
+        // Discard points outside the grid
+        if !self.bounds.contains(&point) {
+            return;
+        } 
+
+        self.add_point_recursive(point, color, 0, max_depth);
+    }
+
+    /// Add a point to the octree recursively. If there are already many points
+    /// in the relevant leaf node, the point may be discarded.
+    ///
+    /// This returns true if the point was added, or false if the point was
+    /// discarded because it didn't fit.
+    fn add_point_recursive(
+            &mut self, 
+            point: Vec3, 
+            color: Vec3, 
+            depth: u8, 
+            max_depth: u8) -> bool {
+        let is_leaf = self.is_leaf();
+        let is_full = self.is_full();
+        if is_leaf && !is_full {
+            // Base case 1: We're at a leaf with some space. just add the point. 
+            self.points.add(point, color);
+            self.count += 1;
+            self.color_sum = self.color_sum + color;
+            return true;
+        } else if is_leaf && is_full && depth < max_depth {
+            // Base case 2: We're at a full leaf. Subdivide this node and
+            // retry the add operation on this node which is now an internal
+            // node. Note that this always produces 8 children
+            self.subdivide();
+            return self.add_point_recursive(point, color, depth, max_depth);
+        } else if is_leaf && is_full && depth == max_depth {
+            // Base case 3: We're at a full leaf but we've hit the depth
+            // limit. Just discard the point to prevent infinite loops
+            return false;
+        } else if !is_leaf {
+            // Recursive case: Find the octant which the point is in, and
+            // insert into the child node
+            let quadrant = self.bounds.find_quadrant(&point);
+            let child = &mut self.children[quadrant]
+                .expect("Unexpected null child");
+            let result = child.add_point_recursive(
+                point, color, depth + 1, max_depth);
+            if result {
+                self.count += 1;
+                self.color_sum = self.color_sum + color;
+            }
+            return result;
+        } else {
+            panic!(
+                "Invalid case: is_leaf: {}, is_full: {}, depth: {}/{}",
+                is_leaf,
+                is_full,
+                depth,
+                max_depth);
+        }
+    }
+
+    /// Subdivide a leaf node into an interior node with 8 children, one
+    /// per octant.
+    fn subdivide(&mut self) {
+        assert!(self.is_leaf(), "can only subdivide leaf nodes");
+        let child_bounds = self.bounds.subdivide();
+        for (i, bounds) in child_bounds.into_iter().enumerate() {
```

Comment:
With the `Vec` implementation, no need for the index, just push!

Also, double check that the bounds are returned in quadrant order (should be the case)